### PR TITLE
NDRS-1049: Drain reactor queue switch

### PR DESCRIPTION
**Pull request by marc-casperlabs**
This prevents fatal errors from being skipped by clearing the `Control` queue and looking for datal events before successfully exiting.

### CODE REVIEW

File: node/src/utils/round_robin.rs [hunk 1]
```diff
@@ -58,6 +61,14 @@ impl<I> QueueState<I> {
         }
     }
 
+    /// Remove all events from a queue.
+    async fn drain(&self) -> Vec<I> {
```
Comment (goral09) [inline node/src/utils/round_robin.rs:65]: Could this return an iterator instead?

File: node/src/reactor.rs [hunk 2]
```diff
@@ -702,6 +702,24 @@ where
...
+                        for event in self.scheduler.drain_queue(QueueKind::Control).await {
+                            if let Some(ctrl_ann) = event.as_control() {
+                                match ctrl_ann {
+                                    ControlAnnouncement::FatalError { file, line, msg } => {
+                                        warn!(%file, line=*line, %msg, "exiting due to fatal error scheduled before reactor completion");
+                                        return ReactorExit::ProcessShouldExit(ExitCode::Abort);
+                                    }
+                                }
+                            } else {
+                                warn!(%event, "found non-control announcement while draining queue")
```
Comment (goral09) [inline node/src/reactor.rs:719]: Won't we end up with many many warns?

**Review by goral09 — APPROVED** [review]

**Comment by goral09** [comment]
@marc-casperlabs Are you planning on merging this?

**Comment by marc-casperlabs** [comment]
This is temporarily going back into draft mode to address a potential issue with the queue's concurrency safety.

**Review by Fraser999 — APPROVED** [review]

File: node/src/utils/round_robin.rs (hunk 1 above)
Comment (marc-casperlabs) [reply node/src/utils/round_robin.rs:65]: An iterator would block the queue permanently while it exists.

```
1. LOCK QUEUE
2. DRAIN ITEMS (<--- iterator would replace this step and would need to hold the lock)
3. UNLOCK QUEUE
```

is what is happening here.

If we returned an iterator, we'd essentially hold the lock as long as it is being iterated, and we would need a guard type to do so. I'm not sure this is worth the extra effort?

File: node/src/reactor.rs (hunk 2 above)
Comment (marc-casperlabs) [reply node/src/reactor.rs:719]: Yes, you're right! I'll downgrade this to a `debug!` message.

File: node/src/utils/round_robin.rs (hunk 1 above)
Comment (goral09) [reply node/src/utils/round_robin.rs:65]: Maybe not, I was just worried that "materializing" all the events into a `Vec` may be problematic.

Comment (marc-casperlabs) [reply node/src/utils/round_robin.rs:65]: They are just being moved. The "draining" happens twice over the lifetime of a node, as in only when we change from initializer to joiner, and from joiner to validator. I think it is fairly unproblematic, and we are aiming to replace this in the mid-term.

**Pull request merged**